File: src/main/go/leetcode/lists/topinterview/LeetCode227.go
package topinterview

func calculate(s string) int {
	if len(s) == 0 {
		return 0
	}
	stack := make([]int, 0)
	sign := '+'
	num := 0
	for i := 0; i < len(s); i++ {
		ch := s[i]
		if IsNumber(ch) {
			num = num*10 + int(ch-'0')
		}
		if !IsNumber(ch) && s[i] != ' ' || i == len(s)-1 {
			switch sign {
			case '+':
				stack = append(stack, num)
				break
			case '-':
				stack = append(stack, -num)
				break
			case '*':
				last := stack[len(stack)-1]
				stack = stack[:len(stack)-1]
				stack = append(stack, last*num)
				break
			case '/':
				last := stack[len(stack)-1]
				stack = stack[:len(stack)-1]
				stack = append(stack, last/num)
				break
			}
			sign = int32(s[i])
			num = 0
		}
	}
	ans := 0
	//意思是最后算加减法, 并且减法转换成加法运算
	for len(stack) > 0 {
		num := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		ans += num
	}
	return ans
}

File: src/main/java/leetcode/lists/topinterview/LeetCode227.java
package leetcode.lists.topinterview;

import java.util.*;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/10/3 18:11
 */
public class LeetCode227 {

    public int calculate(String s) {
        if (s == null || s.length() == 0) {
            return 0;
        }
        char pre = '+';
        Stack<Integer> stack = new Stack<>();
        int num = 0;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (Character.isDigit(c)) {
                num = num * 10 + s.charAt(i) - '0';
            }

            if (!Character.isDigit(c) && c != ' ' || i == s.length() - 1) {
                switch (pre) {
                    case '+' -> stack.push(num);
                    case '-' -> stack.push(-num);
                    case '*' -> stack.push(stack.pop() * num);
                    case '/' -> stack.push(stack.pop() / num);
                }
                pre = c;
                num = 0;
            }
        }
        int ans = 0;
        while (!stack.isEmpty()) {
            ans += stack.pop();
        }
        return ans;
    }
}